File: Client/orgTable/code/ProtocolScene.lua
local descriptor = require "descriptor"

module "ProtocolScene"

SceneObjectStatus = descriptor.def_enum("SceneObjectStatus",
	{
		descriptor.def_enum_value("SOS_STAND", 0),
		descriptor.def_enum_value("SOS_WALK", 2),
	}
)

//  红点标记

RedPointFlag = descriptor.def_enum("RedPointFlag",
	{
		//  公会请求者

		descriptor.def_enum_value("GUILD_REQUESTER", 0),
		//  公会商店

		descriptor.def_enum_value("GUILD_SHOP", 1),
	}
)

ScenePosition = descriptor.def_struct("ScenePosition", 
	{
		descriptor.def_scalar_field("x", descriptor.type_uint32, 0),
		descriptor.def_scalar_field("y", descriptor.type_uint32, 0),
	}
)

SceneDir = descriptor.def_struct("SceneDir", 
	{
		descriptor.def_scalar_field("x", descriptor.type_int16, 0),
		descriptor.def_scalar_field("y", descriptor.type_int16, 0),
		descriptor.def_scalar_field("faceRight", descriptor.type_uint8, 0),
	}
)

SceneNotifyEnterScene = descriptor.def_message("SceneNotifyEnterScene", 500303, 
	{
		descriptor.def_scalar_field("result", descriptor.type_uint32, 0),
		descriptor.def_scalar_field("mapid", descriptor.type_uint32, 0),
		descriptor.def_scalar_field("name", descriptor.type_string, ""),
		descriptor.def_message_field("pos", ScenePosition),
		descriptor.def_message_field("dir", SceneDir),
	}
)

SceneMoveRequire = descriptor.def_message("SceneMoveRequire", 500501, 
	{
		descriptor.def_message_field("pos", ScenePosition),
		descriptor.def_message_field("dir", SceneDir),
	}
)

SceneSyncPlayerMove = descriptor.def_message("SceneSyncPlayerMove", 500502, 
	{
		descriptor.def_scalar_field("id", descriptor.type_uint64, 0),
		descriptor.def_message_field("pos", ScenePosition),
		descriptor.def_message_field("dir", SceneDir),
	}
)

SceneSynSelf = descriptor.def_message("SceneSynSelf", 500601, 
	{
	}
)

SceneSyncSceneObject = descriptor.def_message("SceneSyncSceneObject", 500602, 
	{
	}
)

SceneSyncObjectProperty = descriptor.def_message("SceneSyncObjectProperty", 500603, 
	{
	}
)

SceneDeleteSceneObject = descriptor.def_message("SceneDeleteSceneObject", 500604, 
	{
	}
)

SceneReturnToTown = descriptor.def_message("SceneReturnToTown", 500517, 
	{
	}
)

ScenePlayerChangeSceneReq = descriptor.def_message("ScenePlayerChangeSceneReq", 500503, 
	{
		descriptor.def_scalar_field("curDoorId", descriptor.type_uint32, 0),
		descriptor.def_scalar_field("dstSceneId", descriptor.type_uint32, 0),
		descriptor.def_scalar_field("dstDoorId", descriptor.type_uint32, 0),
	}
)

//  清除公会红点

SceneClearRedPoint = descriptor.def_message("SceneClearRedPoint", 500617, 
	{
		//  红点类型（对应枚举RedPointFlag）

		descriptor.def_scalar_field("flag", descriptor.type_uint32, 0),
	}
)

//  设置用户自定义字段

SceneSetCustomData = descriptor.def_message("SceneSetCustomData", 500620, 
	{
		descriptor.def_scalar_field("data", descriptor.type_uint32, 0),
	}
)

File: Client/orgTable/code/ProtocolScene.cs
using System;
using System.Text;

namespace Protocol
{
	public enum SceneObjectStatus
	{
		SOS_STAND = 0,
		SOS_WALK = 2,
	}

	/// <summary>
	///  红点标记
	/// </summary>
	public enum RedPointFlag
	{
		/// <summary>
		///  公会请求者
		/// </summary>
		GUILD_REQUESTER = 0,
		/// <summary>
		///  公会商店
		/// </summary>
		GUILD_SHOP = 1,
	}

	public class ScenePosition : Protocol.IProtocolStream
	{
		public UInt32 x;
		public UInt32 y;

		#region METHOD

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, x);
				BaseDLL.encode_uint32(buffer, ref pos_, y);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref x);
				BaseDLL.decode_uint32(buffer, ref pos_, ref y);
			}
		#endregion

	}

	public class SceneDir : Protocol.IProtocolStream
	{
		public Int16 x;
		public Int16 y;
		public byte faceRight;

		#region METHOD

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_int16(buffer, ref pos_, x);
				BaseDLL.encode_int16(buffer, ref pos_, y);
				BaseDLL.encode_int8(buffer, ref pos_, faceRight);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_int16(buffer, ref pos_, ref x);
				BaseDLL.decode_int16(buffer, ref pos_, ref y);
				BaseDLL.decode_int8(buffer, ref pos_, ref faceRight);
			}
		#endregion

	}

	[Protocol]
	public class SceneNotifyEnterScene : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500303;
		public UInt32 result;
		public UInt32 mapid;
		public string name;
		public ScenePosition pos = new ScenePosition();
		public SceneDir dir = new SceneDir();

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, result);
				BaseDLL.encode_uint32(buffer, ref pos_, mapid);
				byte[] nameBytes = StringHelper.StringToUTF8Bytes(name);
				BaseDLL.encode_string(buffer, ref pos_, nameBytes, (UInt16)(buffer.Length - pos_));
				pos.encode(buffer, ref pos_);
				dir.encode(buffer, ref pos_);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref result);
				BaseDLL.decode_uint32(buffer, ref pos_, ref mapid);
				UInt16 nameLen = 0;
				BaseDLL.decode_uint16(buffer, ref pos_, ref nameLen);
				byte[] nameBytes = new byte[nameLen];
				for(int i = 0; i < nameLen; i++)
				{
					BaseDLL.decode_int8(buffer, ref pos_, ref nameBytes[i]);
				}
				name = StringHelper.BytesToString(nameBytes);
				pos.decode(buffer, ref pos_);
				dir.decode(buffer, ref pos_);
			}
		#endregion

	}

	[Protocol]
	public class SceneMoveRequire : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500501;
		public ScenePosition pos = new ScenePosition();
		public SceneDir dir = new SceneDir();

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				pos.encode(buffer, ref pos_);
				dir.encode(buffer, ref pos_);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				pos.decode(buffer, ref pos_);
				dir.decode(buffer, ref pos_);
			}
		#endregion

	}

	[Protocol]
	public class SceneSyncPlayerMove : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500502;
		public UInt64 id;
		public ScenePosition pos = new ScenePosition();
		public SceneDir dir = new SceneDir();

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint64(buffer, ref pos_, id);
				pos.encode(buffer, ref pos_);
				dir.encode(buffer, ref pos_);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint64(buffer, ref pos_, ref id);
				pos.decode(buffer, ref pos_);
				dir.decode(buffer, ref pos_);
			}
		#endregion

	}

	[Protocol]
	public class SceneSynSelf : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500601;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
			}

			public void decode(byte[] buffer, ref int pos_)
			{
			}
		#endregion

	}

	[Protocol]
	public class SceneSyncSceneObject : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500602;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
			}

			public void decode(byte[] buffer, ref int pos_)
			{
			}
		#endregion

	}

	[Protocol]
	public class SceneSyncObjectProperty : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500603;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
			}

			public void decode(byte[] buffer, ref int pos_)
			{
			}
		#endregion

	}

	[Protocol]
	public class SceneDeleteSceneObject : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500604;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
			}

			public void decode(byte[] buffer, ref int pos_)
			{
			}
		#endregion

	}

	[Protocol]
	public class SceneReturnToTown : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500517;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
			}

			public void decode(byte[] buffer, ref int pos_)
			{
			}
		#endregion

	}

	[Protocol]
	public class ScenePlayerChangeSceneReq : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500503;
		public UInt32 curDoorId;
		public UInt32 dstSceneId;
		public UInt32 dstDoorId;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, curDoorId);
				BaseDLL.encode_uint32(buffer, ref pos_, dstSceneId);
				BaseDLL.encode_uint32(buffer, ref pos_, dstDoorId);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref curDoorId);
				BaseDLL.decode_uint32(buffer, ref pos_, ref dstSceneId);
				BaseDLL.decode_uint32(buffer, ref pos_, ref dstDoorId);
			}
		#endregion

	}

	/// <summary>
	///  清除公会红点
	/// </summary>
	[Protocol]
	public class SceneClearRedPoint : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500617;
		/// <summary>
		///  红点类型（对应枚举RedPointFlag）
		/// </summary>
		public UInt32 flag;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, flag);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref flag);
			}
		#endregion

	}

	/// <summary>
	///  设置用户自定义字段
	/// </summary>
	[Protocol]
	public class SceneSetCustomData : Protocol.IProtocolStream, Protocol.IGetMsgID
	{
		public const UInt32 MsgID = 500620;
		public UInt32 data;

		#region METHOD
			public UInt32 GetMsgID()
			{
				return MsgID;
			}

			public void encode(byte[] buffer, ref int pos_)
			{
				BaseDLL.encode_uint32(buffer, ref pos_, data);
			}

			public void decode(byte[] buffer, ref int pos_)
			{
				BaseDLL.decode_uint32(buffer, ref pos_, ref data);
			}
		#endregion

	}

}
